homeassistant/components/melnor/sensor.py

```diff
@@ -0,0 +1,128 @@
+"""Support for Melnor RainCloud sprinkler water timer."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+from typing import Any
+
+from melnor_bluetooth.device import Device
+
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import PERCENTAGE, SIGNAL_STRENGTH_DECIBELS_MILLIWATT
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity import EntityCategory
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.typing import StateType
+
+from .const import DOMAIN
+from .models import MelnorBluetoothBaseEntity, MelnorDataUpdateCoordinator
+
+
+@dataclass
+class MelnorSensorEntityDescriptionMixin:
+    """Mixin for required keys."""
+
+    state_fn: Callable[[Device], Any]
+
+
+@dataclass
+class MelnorSensorEntityDescription(
+    SensorEntityDescription, MelnorSensorEntityDescriptionMixin
+):
+    """Describes Melnor sensor entity."""
+
+    icon_fn: Callable[[Device], str] | None = None
+
+
+def _rssi_icon(device: Device) -> str:
+    """Return a representative icon based on signal strength."""
+
+    if device.rssi >= -20:
+        return "mdi:wifi-strength-4"
+    if device.rssi >= -40:
+        return "mdi:wifi-strength-3"
+    if device.rssi >= -60:
+        return "mdi:wifi-strength-2"
+    if device.rssi >= -80:
+        return "mdi:wifi-strength-1"
+    return "mdi:wifi-strength-outline"
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_devices: AddEntitiesCallback,
+) -> None:
+    """Set up the sensor platform."""
+
+    coordinator: MelnorDataUpdateCoordinator = hass.data[DOMAIN][config_entry.entry_id]
+
+    sensors: list[MelnorSensorEntityDescription] = [
+        MelnorSensorEntityDescription(
+            device_class=SensorDeviceClass.BATTERY,
+            entity_category=EntityCategory.DIAGNOSTIC,
+            key="battery",
+            name="Battery",
+            native_unit_of_measurement=PERCENTAGE,
+            state_class=SensorStateClass.MEASUREMENT,
+            state_fn=lambda device: device.battery_level,
+        ),
+        MelnorSensorEntityDescription(
+            device_class=SensorDeviceClass.SIGNAL_STRENGTH,
+            entity_category=EntityCategory.DIAGNOSTIC,
+            entity_registry_enabled_default=False,
+            icon_fn=_rssi_icon,
```

Comment:
In our developer documentation we specify that the icon for the device class generally shouldn't be overridden

Reply:
Generally, that makes sense. But for this specific case I think it could be acceptable to customize the icon. It's simply a better user experience for the less-technical users to see an icon that represents the quality of the signal strength being reported. I got the idea from a couple of other integrations in the codebase doing the same thing. The default icon being used by the `signal_strength` device class doesn't have other strength representations otherwise I'd have stuck with the default.

https://github.com/home-assistant/core/blob/fc9a0ba46bbbf8f7d191569c5a81d74c9dd732ff/homeassistant/components/advantage_air/sensor.py#L137-L148

https://github.com/home-assistant/core/blob/41f63839571643b09b0cd7ea55375ea6c4f3abe9/homeassistant/components/huawei_lte/sensor.py#L219-L232

Reply:
This is a frontend concern so it really should be a PR to https://github.com/home-assistant/frontend

Reply:
So is the implementation I have now a complete no-go for this PR?

Reply:
Let's drop it from this PR.  The above examples are legacy.  We don't want to solve this over and over again.  It really should be done once on the frontend.

Reply:
done